[pls do not squash :3] Allow limiting object observers (#13987)
minetest/minetest: doc/lua_api.md

    * only properties listed in the table will be changed
    * see the 'Object properties' section for details
* `get_properties()`: returns a table of all object properties
* `set_observers(observers)`: sets observers (players this object is sent to)
    * If `observers` is `nil` (or nothing), the object's observers are "unmanaged":
      The object is sent to all players (in proximity). This is the default.
    * `observers` is a "set" of player names: `{[player name] = true, [other player name] = true, ...}`
    * If players are managed, they always need to have themselves as observers.
    * Attachments:
        * If an object's observers are managed, the observers of all children need to be managed too.
        * The observers of children need to be a subset of the observers of parents.
* `get_observers()`:

>     * returns nothing if the observers are unmanaged

Reviewer: IMO we shouldn't differentiate nil and nothing in the api docs or usage
```suggestion
    * returns `nil` if the observers are unmanaged
```

Author: Should I make the code actually push a `nil` then? I might be pedantic here, but documenting this as `return nil` when it's actually `return` feels wrong to me; there are cases where it could matter.

Reviewer: it should push `nil` IMO

Author: Done. Also made it throw an error for invalid object refs to avoid surprises.

    * only properties listed in the table will be changed
    * see the 'Object properties' section for details
* `get_properties()`: returns a table of all object properties
* `set_observers(observers)`: sets observers (players this object is sent to)
    * If `observers` is `nil`, the object's observers are "unmanaged":

>       The object is sent to all players (in proximity). This is the default.

Reviewer: ```suggestion
      The object is sent to all players as governed by server settings. This is the default.
```

    * `observers` is a "set" of player names: `{[player name] = true, [other player name] = true, ...}`
        * A player is an observer if and only if the player's name is a key in the `observers` table.
        * The values are irrelevant. To prevent the mistake of expecting `[player name] = false`

>           to "uninstall" a player as an observer, they must be truthy.

Reviewer: I would simply say that they must be true, don't want to give modders too much leeway.
(Unrelated to what the engine code actually does)

    * If players are managed, they always need to have themselves as observers.
    * Attachments:
        * If an object's observers are managed, the observers of all children need to be managed too.

>         * The observers of children need to be a subset of the observers of parents.

Reviewer: unanswered question: object (un)loading is not touched by observers, right?

Reviewer: ^ don't see this answered yet

* `get_bone_overrides()`: returns all bone overrides as table `{[bonename] = override, ...}`
* `set_properties(object property table)`
* `get_properties()`: returns a table of all object properties

> * `set_observers(observers)`: sets observers (players this object is sent to)

Author: there are two reasons why i want to be explicit here:

* there's more than visibility to "observability". there's sound for example; there may be an invisible object with sounds attached. (now that i think of it, i doubt that this is currently handled correctly; handling it correctly will probably have to be based on #14342)
* one specific reason this is interesting for modders is because it can be used to truly hide objects from players, even if cheat clients are involved.

* `get_bone_overrides()`: returns all bone overrides as table `{[bonename] = override, ...}`
* `set_properties(object property table)`
* `get_properties()`: returns a table of all object properties
* `set_observers(observers)`: sets observers (players this object is sent to)
    * If `observers` is `nil`, the object's observers are "unmanaged":
      The object is sent to all players as governed by server settings. This is the default.
    * `observers` is a "set" of player names: `{[player name] = true, [other player name] = true, ...}`

>         * Since this is a set, the values need to be `true`.

Author: I'm a fan of consistency, but I think the term "flag field" would be confusing here, especially since we support some additional specifier formats there (strings). I also find the term less intuitive.

I can properly define the term. To me "set" is sort of "extended" Lua lingo, just like you'd probably call a table with keys 1, 2, ..., n a "list". But it seems this notion is not that widespread so I should probably clarify it.

    * If players are managed, they always need to have themselves as observers.
    * Attachments: The "final observers" are the intersection of all observer sets
      of all ancestors of an object and the object itself.
      The intersection of an unmanaged observer set with a managed one is the managed one,

>       the intersection of two unmanaged observer sets is unmanaged.

Author: Agree, that wording was convoluted, reworded:

> The *effective observers* of an object are made up of all players who can observe the object *and* are also effective observers of its parent object (if there is one).

* `get_bone_overrides()`: returns all bone overrides as table `{[bonename] = override, ...}`
* `set_properties(object property table)`
* `get_properties()`: returns a table of all object properties
* `set_observers(observers)`: sets observers (players this object is sent to)
    * If `observers` is `nil`, the object's observers are "unmanaged":
      The object is sent to all players as governed by server settings. This is the default.
    * `observers` is a "set" of player names: `{[player name] = true, [other player name] = true, ...}`
        * A set is a table where the keys are the elements of the set (in this case, player names)
          and the values are all `true`.
    * Attachments: The *effective observers* of an object are made up of
      all players who can observe the object *and* are also effective observers
      of its parent object (if there is one).
	* Players are automatically added to their own observer sets.

> 	  Players **must** effectively observe themselves.

Author: Yes, if they don't abide by this they may get the mod error about players not observing themselves.
